### DISCUSSION TITLE
Way to reference column (or default) without triggering binder error

### DISCUSSION THREAD

**Discussion by zenazn** [duckdb/duckdb#9052] [Ideas]
I have a set of parquet files that were generated at various points in time. We've evolved the schema over time, adding columns, some of which have non-null "default values."

I'd like to write a query that works across all versions of this file. Perhaps concretely it generates a `VIEW` that abstracts over the data differences in these files, by filling in the default value if the column isn't present (but I'm open to other solutions). One solution that feels good to me might look like:

``` sql
SELECT
  "column1"
  try_struct_extract(foo, 'column2', 'default value') as "column2",
  try_struct_extract(foo, 'column3') as "column3", -- with no default, NULL is assumed
FROM foo
```

However, all of the ways I can think of to do this trigger a binder error. Due to implementation details of our architecture I'm happiest with options that can be performed "non-interactively" — that is, given a list of columns and their default values, I can statically generate a single SQL query that will work across all files. I know that with a few `DESCRIBE` back-and-forths I can generate an appropriate query, but that's not as good a fit for my use case.

My current best workaround is to use the "auto-incrementing duplicate column names" from https://duckdb.org/2022/05/04/friendlier-sql.html:

```
WITH __defaults AS (
  SELECT 'default value' AS "column2", NULL as "column3"
), __merged AS (
  SELECT * FROM foo CROSS JOIN __defaults
)
SELECT COLUMNS(c -> c NOT LIKE '%:1') FROM __merged
```

But this approach seems dissatisfying (e.g., it makes some assumptions about what a "valid" column name looks like)

**Comment by l1t1** [comment]
use unpivot to get the column name and right join to full column list, then pivot it.

**Comment by l1t1** [comment]
use unpivot to get the column name and right join to full column list, then pivot it.

**Comment by l1t1** [comment]
https://duckdb.org/docs/data/parquet/tips#use-union_by_name-when-loading-files-with-different-schemas

**Comment by l1t1** [comment]
```
D copy(select 1 a) to 't1.parquet';
D copy(select 1 a, 2 b) to 't2.parquet';
D
D SELECT * FROM read_parquet('t*.parquet', union_by_name=True);
┌───────┬───────┐
│   a   │   b   │
│ int32 │ int32 │
├───────┼───────┤
│     1 │       │
│     1 │     2 │
└───────┴───────┘
D SELECT * FROM read_parquet('t*.parquet', union_by_name=True, filename=True);
┌───────┬───────┬────────────┐
│   a   │   b   │  filename  │
│ int32 │ int32 │  varchar   │
├───────┼───────┼────────────┤
│     1 │       │ t1.parquet │
│     1 │     2 │ t2.parquet │
└───────┴───────┴────────────┘
D LECT * FROM read_parquet('t*.parquet', union_by_name=True, filename=True) where filename='t2.parquet';
┌───────┬───────┬────────────┐
│   a   │   b   │  filename  │
│ int32 │ int32 │  varchar   │
├───────┼───────┼────────────┤
│     1 │     2 │ t2.parquet │
└───────┴───────┴────────────┘
D SELECT * FROM read_parquet('t*.parquet', union_by_name=True, filename=True) where filename='t1.parquet';
┌───────┬───────┬────────────┐
│   a   │   b   │  filename  │
│ int32 │ int32 │  varchar   │
├───────┼───────┼────────────┤
│     1 │       │ t1.parquet │
```

**Comment by l1t1** [comment]
```
D SELECT * FROM read_parquet('t1.parquet') union by name SELECT * FROM read_parquet('t2.parquet');
┌───────┬───────┐
│   a   │   b   │
│ int32 │ int32 │
├───────┼───────┤
│     1 │       │
│     1 │     2 │
└───────┴───────┘
D SELECT * FROM read_parquet('t1.parquet') where 1=2 union by name SELECT * FROM read_parquet('t2.parquet');
┌───────┬───────┐
│   a   │   b   │
│ int32 │ int32 │
├───────┼───────┤
│     1 │     2 │
└───────┴───────┘
D SELECT * FROM read_parquet('t1.parquet') union by name SELECT * FROM read_parquet('t2.parquet') where 1=2;
┌───────┬───────┐
│   a   │   b   │
│ int32 │ int32 │
├───────┼───────┤
│     1 │       │
```
